### ISSUE TITLE
What do the 17 lines with no language and skill level ‘0’ mean?

### ISSUE THREAD

**Issue by mramosch** [Tatoeba/tatoeba2#2644]
User skill level per language

Filename
user_languages.tar.bz2
![IMG_2565](https://user-images.githubusercontent.com/725445/110954234-8d832c00-8348-11eb-80a8-cdf37d8fad07.jpeg)

**Comment by mramosch** [comment]
![IMG_2567](https://user-images.githubusercontent.com/725445/110964779-d68cad80-8353-11eb-9265-efe9207b0f2d.jpeg)

**Comment by mramosch** [comment]
Sorted by username:

———
———
———


![IMG_2569](https://user-images.githubusercontent.com/725445/110964939-063bb580-8354-11eb-87fb-f80fc517f031.jpeg)


![IMG_2570](https://user-images.githubusercontent.com/725445/110965152-37b48100-8354-11eb-8139-129bc21170a1.jpeg)


![IMG_2571](https://user-images.githubusercontent.com/725445/110965292-5581e600-8354-11eb-8b79-8d8e09f7b805.png)


![IMG_2572](https://user-images.githubusercontent.com/725445/110965365-66325c00-8354-11eb-9ca8-9efe46da77bc.png)


![IMG_2573](https://user-images.githubusercontent.com/725445/110965436-78ac9580-8354-11eb-9a42-d501e8b31a8d.jpeg)

**Comment by jiru** [comment]
Thank you for the report. :slightly_smiling_face: Among all the things you pointed out, some are actual bugs while some are not.

Non-bugs:
* Integers are actual usernames, for example: https://tatoeba.org/user/profile/6300669966. Your spreadsheet right-aligns them because it guesses they are integers, but they must to be treated as strings.
* "SOS 5" is what your spreadsheet displays for the username 5SOS for some reason. There is not much we can do about it.

Bugs:
* Empty language codes are a bug, not sure how these got it.

Confusing behavior:
* `\N` means "undefined". This is when there are no stars displayed at all for that language in the profile page. This probably needs to be changed or explained on the Downloads page.
* The strange line about Saru dialect is a running line preceded by a backslash. This is how mysql encodes the `\r` character: it writes it as-is and adds a backslash right before it. We probably want to get rid of this using `mysql --batch` (like in #2360). For now you can safely ignore such lines if you don’t care about what users wrote in the "details" field.

**Comment by mramosch** [comment]
I am not referring to \N as an undefined language but rather to their appearance as user name!

Is this block of two lines

 \;;
saru kotan...;;

one record or are these two records? There is no language code, however...

The parser assumes that the first comma separated value in a new line is the language code with 3 letters followed by a semicolon. When you’re getting weird stuff like this the parser would have to do a lot more work just to secure against a few strange records...

**Comment by jiru** [comment]
> I am not referring to \N as an undefined language but rather to their appearance as user name!

I don’t think we need to care so much about appearance of the contents of the export files because they are meant to be parsed by programs or read by programmers.

> Is this block of two lines
> 
> ;;
> saru kotan...;;
> 
> one record or are these two records? There is no language code, however...

The block of two lines you are quoting does not come from the original CSV file. I can tell because it is using semi-colons. You need to work with the original CSV file that uses tabs as column separator if you want to get it parsed correctly.

**Comment by mramosch** [comment]
Well, I am using them to be parsed by a program, so every little inconsistency in the file needs an even more sophisticated parser, which wouldn’t be necessary if the file format were ‘valid’ according to the description of the file’s fields on the download page.

I guess it’s much easier to implement user-input gate-keepers on the server-side to guarantee a consistent export file format than guessing on the client-side what could be off-specification.

e.g. you said that integers are valid user names. Well, I started out with the first on the list (user: 1) and guess what? It’s not a valid user name...

That means, if I find a single hick-up in the export file I can’t trust it at all. And that would be a nightmare for weekly updates in an application...

This is no criticism or hate against you guys ;-) But you always ask to provide you with our use-cases and real world scenarios!

**Comment by Yorwba** [comment]
> every little inconsistency in the file needs an even more sophisticated parser

Here's how I've been parsing Tatoeba exports for years. It's not particularly sophisticated, but maybe it will help you:

https://github.com/Yorwba/alphabet-soup/blob/482f6e12babf69c31b23f89c8d2ea80a409802e6/tatoeba_data.py#L23-L65

> you said that integers are valid user names. Well, I started out with the first on the list (user: 1) and guess what? It’s not a valid user name...

That's because the user is [0001](https://tatoeba.org/cmn/user/profile/0001), not 1. You really shouldn't use a grid calculator to open these files. I recommend a plain-text editor or some other simple tool that displays the contents of the file as they actually are.

**Comment by mramosch** [comment]
Thanks, for the link ;-) I’ll have a look at it...

I DO have the download files opened in a text editor, too.

But to get a feel for the structure (sorting, re-ordering and deleting columns) it was much easier with a grid view ;-)
And then exporting the resulting data as CVS and using it for the parser...

But I get your point, I’ll give it some thoughts...

What about the user name: \N

Knows two languages but doesn’t appear on the Tatoeba user page

![IMG_2581](https://user-images.githubusercontent.com/725445/111072068-48daca80-84d9-11eb-9459-56c5c6c840d3.jpeg)


![IMG_2585](https://user-images.githubusercontent.com/725445/111072095-614ae500-84d9-11eb-87ae-894e174daae2.jpeg)

**Comment by jiru** [comment]
> What about the user name: \N

This is a bug. This language skill belongs to a user that does not exist. I guess that happened as a result of user deletion.

**Comment by mramosch** [comment]
OK.

So just to be on the same page -

If I take the original download CSV file, can I take it for granted that, even if a new line doesn’t necessarily indicate the start of a new data-record, that the structure of 4 fields is maintained, hence that a new record starts after every 4 tabs?

———

**Comment by jiru** [comment]
Not exactly. A new record does not start after every 4 tabs; it starts after a new line (unless that new line is preceded by a backslash). Let me highlight the part of the CSV file that requires special parsing. You should be able to see it as well if you open the CSV file with a text editor.

```
ain	5	reviewsmonkey	
ain	5	sarah9425	
ain	4	killya	
ain	1	Silja	shikot kotan'itak・シコッ　コタン　イタㇰ・千歳方言・Chitose dialect・\
saru kotan'itak・サル　コタン　イタㇰ・沙流方言・Saru dialect
ain	0	MomijiAkira	
ain	0	protondonor	
```

The line about Silja contains 4 fields, but the 4th field is a text that happens to include a newline character. However, we already use the newline character to separate records. So how to include a newline in the 4th field without creating a new record? Using the backslash. The backslash means "please read the next character as a regular character and not as a record separator". This is what the Python code Yorwba mentioned is doing. For comparison you can check how the details about Ainu language look on [Silja](https://tatoeba.org/user/profile/Silja)’s profile. In particular you will see that there is no backslash there. (And the newline is ignored, but that’s a different problem.)

I agree that it would be better to use a simpler encoding system. My suggestion above about `mysql --batch` would result into having newlines encoded as `\n`. Until this is done, please bear with the current state of affairs.

**Comment by mramosch** [comment]
Where is the newline character in the 4th field coming from in the first place?

If there were no linebreaks allowed in the 4th field then the whole escape sequence dance would be avoidable without any exception to the rules. Is paragraphing or linebreaking really a necessity in a little detail description field?

**Comment by jiru** [comment]
It is there because it’s possible to insert newlines there. That field is a textarea. If it’s a problem for you, you are welcome to submit a pull request. :slightly_smiling_face:

**Comment by ckjpn** [comment]
\N for the owner's username means "no owner" (orphan sentences)

**Comment by mramosch** [comment]
But there are only 2 of them and they have languages attached to them. That doesn’t make sense...

**Comment by ckjpn** [comment]
I thought you meant the 4th field in the "sentences_detailed.csv" file.

Like these sentences.

```
8577511	jpn	私のプロフィールの写真は美しいですか。	\N	2020-02-25 21:49:51	2020-03-04 04:33:21
8603503	jpn	トムは床の上に血が付いているのを見た。	\N	2020-03-09 02:41:42	2020-09-18 11:15:41
8642761	jpn	トムは鹿をはね、車がダメになった。	\N	2020-03-29 08:36:21	2020-09-18 11:08:20
8950559	jpn	少年と女子は読みます。	\N	2020-07-31 02:33:36	2020-11-13 22:41:44
9366206	jpn	トムは吃音するの人。	\N	2020-11-24 09:02:12	2020-11-24 21:23:08

```
